internal/healthchecks/builtin.go

```diff
@@ -0,0 +1,170 @@
+package healthchecks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AreObjectsHealthy checks if the given resources are healthy.
+// It returns a nil error if all the resources are healthy, if any resource is not healthy, the error will
+// contain the GVK + namespace/resourceName and the error message of each unhealthy resource.
+//
+// The current list of supported resources is:
+// - Deployments
+// - StatefulSets
+// - DaemonSets
+// - ReplicaSets
+// - Pods
+// - APIServices
+// - CustomResourceDefinitions
+//
+// If the resource is not supported, it is assumed to be healthy.
+func AreObjectsHealthy(ctx context.Context, client client.Client, objects []client.Object) error {
+	var gvkErrors []error
+
+	for _, object := range objects {
+		// Convert the object to its proper type, deepCopy it to avoid mutating the original object.
+		obj, err := runtime.ObjectConvertor(client.Scheme()).ConvertToVersion(object.DeepCopyObject(), object.GetObjectKind().GroupVersionKind().GroupVersion())
+		if err != nil {
+			gvkErrors = appendResourceError(gvkErrors, object, err.Error())
+			continue
+		}
+
+		objectKey := types.NamespacedName{
+			Name:      object.GetName(),
+			Namespace: object.GetNamespace(),
+		}
+
+		switch obj := obj.(type) {
+		case *appsv1.Deployment:
+			if err := client.Get(ctx, objectKey, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			// Check if the deployment is available.
+			conditionExists := false
+			for _, condition := range obj.Status.Conditions {
+				if condition.Type == appsv1.DeploymentAvailable {
+					if condition.Status != "True" {
+						gvkErrors = appendResourceError(gvkErrors, obj, condition.Message)
+					}
+					conditionExists = true
+					break
+				}
+			}
+			if conditionExists {
+				continue
+			}
+			gvkErrors = appendResourceError(gvkErrors, obj, "DeploymentAvailable condition not found")
+		case *appsv1.StatefulSet:
+			if err := client.Get(ctx, objectKey, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.ReadyReplicas != obj.Status.Replicas {
```

Comment:
Global comment: It seems like Helm's ready checking logic is much more [thorough](https://github.com/helm/helm/blob/e7bb860d9a32e8739c944b8e7b7f7031d752411a/pkg/kube/ready.go#L357-L410). It seems like our code could return false positives and flap in a variety of common situations. Should we re-use helm's code somehow?

Reply:
I will expand the logic.

Reply:
done, please take a look.